src/transformers/models/gpt_neox/modeling_gpt_neox.py

```diff
@@ -204,43 +260,38 @@ def forward(
             value = torch.cat((past_value, value), dim=-2)
         present = (key, value) if use_cache else None
 
-        # Compute attention
-        attn_output, attn_weights = self._attn(query, key, value, attention_mask, head_mask)
-
-        # Reshape outputs
-        attn_output = self._merge_heads(attn_output, self.num_attention_heads, self.head_size)
-        attn_output = self.dense(attn_output)
+        # GPT-neo-X casts query and key in fp32 to apply rotary embedding in full precision
+        target_dtype = value.dtype
+        if query.dtype != target_dtype:
+            query = query.to(target_dtype)
+        if key.dtype != target_dtype:
+            key = key.to(target_dtype)
```

Comment:
This has been ignored in the base attn implementation but I don't see a reason why.

Reply:
This led to some strange behaviour so casts are back to the respective attn implementations.